File: benchmarks/Java/src/deltablue/ScaleConstraint.java
/*
 * This benchmark is derived from Mario Wolczko's Java and Smalltalk version of
 * DeltaBlue.
 *
 * It is modified to use the SOM class library and Java 8 features.
 * License details:
 *   http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html
 */
package deltablue;

import deltablue.Strength.Sym;
import som.ForEachInterface;

// I relate two variables by the linear scaling relationship: "v2 =
// (v1 * scale) + offset". Either v1 or v2 may be changed to maintain
// this relationship but the scale factor and offset are considered
// read-only.
final class ScaleConstraint extends BinaryConstraint {

  protected final Variable scale;  // scale factor input variable
  protected final Variable offset; // offset input variable

  ScaleConstraint(final Variable src, final Variable scale,
      final Variable offset, final Variable dest, final Sym strength,
      final Planner planner) {
    super(src, dest, strength, planner);
    this.scale = scale;
    this.offset = offset;
    addConstraint(planner);
  }

  // Add myself to the constraint graph.
  @Override
  public void addToGraph() {
    v1.addConstraint(this);
    v2.addConstraint(this);
    scale.addConstraint(this);
    offset.addConstraint(this);
    direction = null;
  }

  // Remove myself from the constraint graph.
  @Override
  public void removeFromGraph() {
    if (v1 != null) { v1.removeConstraint(this); }
    if (v2 != null) { v2.removeConstraint(this); }
    if (scale  != null) { scale.removeConstraint(this); }
    if (offset != null) { offset.removeConstraint(this); }
    direction = null;
  }

  // Enforce this constraint. Assume that it is satisfied.
  @Override
  public void execute() {
    if (direction == Direction.FORWARD) {
      v2.setValue(v1.getValue() * scale.getValue() + offset.getValue());
    } else {
      v1.setValue((v2.getValue() - offset.getValue()) / scale.getValue());
    }
  }

  @Override
  public void inputsDo(final ForEachInterface<Variable> fn) {
    if (direction == Direction.FORWARD) {
      fn.apply(v1);
      fn.apply(scale);
      fn.apply(offset);
    } else {
      fn.apply(v2);
      fn.apply(scale);
      fn.apply(offset);
    }
  }

  // Calculate the walkabout strength, the stay flag, and, if it is
  // 'stay', the value for the current output of this
  // constraint. Assume this constraint is satisfied.
  @Override
  public void recalculate() {
    Variable in;
    Variable out;

    if (direction == Direction.FORWARD) {
      in  = v1; out = v2;
    } else {
      out = v1; in  = v2;
    }

    out.setWalkStrength(strength.weakest(in.getWalkStrength()));
    out.setStay(in.getStay() && scale.getStay() && offset.getStay());
    if (out.getStay()) {
      execute(); // stay optimization
    }
  }
}

File: benchmarks/C++/src/deltablue/ScaleConstraint.cpp
#include "ScaleConstraint.h"

namespace deltablue {

    ScaleConstraint::ScaleConstraint(shared_ptr<Variable> src, shared_ptr<Variable> scale, shared_ptr<Variable> offset,
            shared_ptr<Variable> dest, shared_ptr<Strength::Sym> strength, shared_ptr<Planner> planner)
        : BinaryConstraint(src, dest, strength/*, planner*/) {
        _strength = Strength::of(strength);
        _v1 = src;
        _v2 = dest;
        _direction = NONE;
        _scale = scale;
        _offset = offset;

        //addConstraint(planner);
    }

    void ScaleConstraint::addToGraph() {
        _v1->addConstraint(shared_from_this());
        _v2->addConstraint(shared_from_this());
        _scale->addConstraint(shared_from_this());
        _offset->addConstraint(shared_from_this());
        _direction = NONE;
    }

    void ScaleConstraint::removeFromGraph() {
        if (_v1 != nullptr) 
            _v1->removeConstraint(shared_from_this());
        if (_v2 != nullptr) 
            _v2->removeConstraint(shared_from_this());
        if (_scale != nullptr) 
            _scale->removeConstraint(shared_from_this());
        if (_offset != nullptr) 
            _offset->removeConstraint(shared_from_this());
        _direction = NONE;
    }

    void ScaleConstraint::execute() {
        if (_direction == FORWARD) {
        _v2->setValue(_v1->getValue() * _scale->getValue() + _offset->getValue());
        } else {
        _v1->setValue((_v2->getValue() - _offset->getValue()) / _scale->getValue());
        }
    }

    void ScaleConstraint::inputsDo(function<void(shared_ptr<Variable>)> fn) {
        if (_direction == FORWARD) {
            fn(_v1);
            fn(_scale);
            fn(_offset);
        } else {
            fn(_v2);
            fn(_scale);
            fn(_offset);
        }
    }

    void ScaleConstraint::recalculate() {
        shared_ptr<Variable> in;
        shared_ptr<Variable> out;

        if (_direction == FORWARD) {
            in = _v1;
            out = _v2;
        } else {
            out = _v1;
            in = _v2;
        }

        out->setWalkStrength(_strength->weakest(in->getWalkStrength()));
        out->setStay(in->getStay() && _scale->getStay() && _offset->getStay());
        if (out->getStay()) {
            execute(); // stay optimization
        }
    }
}